ledger/complete/ledger.go

```diff
@@ -229,10 +266,44 @@ func (l *Ledger) Set(update *ledger.Update) (newState ledger.State, trieUpdate *
 
 	state := update.State()
 	l.logger.Info().Hex("from", state[:]).
-		Hex("to", newRootHash[:]).
+		Hex("to", newState[:]).
 		Int("update_size", update.Size()).
 		Msg("ledger updated")
-	return ledger.State(newRootHash), trieUpdate, nil
+	return newState, trieUpdate, nil
+}
+
+func (l *Ledger) set(trieUpdate *ledger.TrieUpdate) (newState ledger.State, err error) {
+
+	resultCh := make(chan error)
+
+	if l.trieUpdateCh == nil {
+		go func() {
+			_, _, err := l.wal.RecordUpdate(trieUpdate)
+			resultCh <- err
+		}()
+	} else {
+		doneCh := make(chan struct{})
```

Comment:
> This means, if ledger.Set is called concurrently by two calls, one of the call will be blocked waiting until the first call is completed.

No, that's not exactly how it works.  

When `Ledger.Set` is called concurrently by two calls, both send `trieUpdate` message to `Compactor` non-blocking, and create new tries from updates concurrently.  What is blocking is `Compactor` can only process one `trieUpdate` at a time, so the second `Ledger.Set` waits for `Compactor` to process its message in turn.  

In other words, two `Ledger.Set` create new tries concurrently while updating WAL and adding new trie to ledger state are in sequence.

Reply:
> IMO, the behavior of a RWMutex lock is easier than the doneCh to understand, even though both can be used to ensure concurrent-safety.

I understand what you're saying but adding RWMutex doesn't help because we need the extra channel based on discussions @m4ksio, you, and I had on Friday morning.

Based Friday's meeting, I replaced `doneCh` with `trieCh` to send new trie from `Ledger.Set` to `Compactor`.  It is safer to use a FIFO queue to store tries for checkpointing in `Compactor` than taking a snapshot of main ledger state.    

With this change, `ledger.Set` should communicate new trie to `Compactor` using the third channel.

One alternative approach to use 2 channels instead of 3 channels is:
- sending both `trieUpdate` and new trie from `Ledger.Set` to `Compactor` in first channel
- waiting for WAL update result in second channel

However, the alternative approach would take longer for updates so it isn't worthwhile.

Reply:
Yes, you are right. Adding the trie update to forest could be concurrent. But the problem is it’s not safe. TrieUpdates could have dependency, and has to be applied sequentially if there is.  For instance, if a block has two trieUpdates, the second update might build on a node created by the first update, adding them concurrently would fail. 
Concurrent update could only work if the blocks of the updates are from different forks, and all the updates they depend on has been added to the forest.
We’ve been running this code for a long time, and never run into problem, is it because concurrent update actually worked? We actually concurrently adding updates to the forest? That’s why I checked the code and see whether Ledger.set could be called concurrently. Turns out, it’s always called sequentially, which means the concurrent update case never happens, and we never run into problem. 

In other words, the ledger set is not safe to be called concurrently with two updates. Ledger has no protocol level knowledge, has no visibility of the dependencies between two updates, so it’s only safe to be called sequentially for now, which is what we do currently.

Reply:
> Adding the trie update to forest could be concurrent. But the problem is it’s not safe. 

No that is incorrect.

In this PR [creating new trie from update](https://github.com/onflow/flow-go/blob/fd43fd5b7c8af9dfbc10030d9634a5ca66e94de8/ledger/complete/ledger.go#L254) is concurrent.  However, [adding the new trie to forest](https://github.com/onflow/flow-go/blob/fd43fd5b7c8af9dfbc10030d9634a5ca66e94de8/ledger/complete/ledger.go#L264) is sequential (by use of channels).

Reply:
>  For instance, if a block has two trieUpdates, the second update might build on a node created by the first update, adding them concurrently would fail.

No it won't fail because trie updates are in sequence for a block, so this case is handled properly.

> Concurrent update could only work if the blocks of the updates are from different forks, and all the updates they depend on has been added to the forest.

Yes, @m4ksio confirmed in #flow-execution that concurrent trie updates only happen if the block forks happened.